### PR TITLE
[SPARK-27255][SQL] Report error when illegal expressions are hosted by a plan operator.

### PR DESCRIPTION
**Pull request by dilipbiswal**
## What changes were proposed in this pull request?
In the PR, we raise an AnalysisError when we detect the presense of aggregate expressions in where clause. Here is the problem description from the JIRA.

Aggregate functions should not be allowed in WHERE clause. But Spark SQL throws an exception when generating codes. It is supposed to throw an exception during parsing or analyzing.

Here is an example:
```
val df = spark.sql("select * from t where sum(ta) > 0")
df.explain(true)
df.show()
```
Resulting exception: 
```
Exception in thread "main" java.lang.UnsupportedOperationException: Cannot generate code for expression: sum(cast(input[0, int, false] as bigint))
	at org.apache.spark.sql.catalyst.expressions.Unevaluable.doGenCode(Expression.scala:291)
	at org.apache.spark.sql.catalyst.expressions.Unevaluable.doGenCode$(Expression.scala:290)
	at org.apache.spark.sql.catalyst.expressions.aggregate.AggregateExpression.doGenCode(interfaces.scala:87)
	at org.apache.spark.sql.catalyst.expressions.Expression.$anonfun$genCode$3(Expression.scala:138)
	at scala.Option.getOrElse(Option.scala:138)
```
Checked the behaviour of other database and all of them return an exception:
**Postgress**
```
select * from foo where max(c1) > 0;
Error
ERROR: aggregate functions are not allowed in WHERE Position: 25
```
**DB2**
```
db2 => select * from foo where max(c1) > 0;
SQL0120N  Invalid use of an aggregate function or OLAP function.  
```
**Oracle**
```
select * from foo where max(c1) > 0;
ORA-00934: group function is not allowed here
```
**MySql**
```
select * from foo where max(c1) > 0;
Invalid use of group function
```

**Update**
This PR has been enhanced to report error when expressions such as Aggregate, Window, Generate are hosted by operators where they are invalid. 
## How was this patch tested?
Added tests in AnalysisErrorSuite and group-by.sql

### CODE REVIEW

**Comment by dilipbiswal** [comment]
cc @cloud-fan @maropu @chakravarthiT

**Comment by maropu** [comment]
To just make sure this is a right approach, one question; `CheckAnalysis` is the best place to fix this issue? For example, analyzer itself throws an exception when filter has generators?

```
scala> sql("select * from x where explode(array(1, 2, 3)) < id").show
org.apache.spark.sql.AnalysisException: Generators are not supported outside the SELECT clause, but got: 'Filter (explode(array(1, 2, 3)) < id#10L);
  at org.apache.spark.sql.catalyst.analysis.Analyzer$ExtractGenerator$$anonfun$apply$17.applyOrElse(Analyzer.scala:1796)
  at org.apache.spark.sql.catalyst.analysis.Analyzer$ExtractGenerator$$anonfun$apply$17.applyOrElse(Analyzer.scala:1751)
  at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.$anonfun$resolveOperatorsUp$3(AnalysisHelper.scala:90)
  at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:72)
```

**Comment by dilipbiswal** [comment]
@maropu In my understanding, the goal was to handle most of semantic errors in checkAnalysis. In cases when its not do-able, they are handled in analyzer itself. Thats my understanding. What do you think Wenchen ?

**Comment by cloud-fan** [comment]
I vaguely remember we did a similar check for window functions, how was that done? cc @rednaxelafx

**Comment by dilipbiswal** [comment]
@cloud-fan I checked the code. The validation for presence of window expression(s) in filter and having is done in analyzer in rule `ExtractWindowExpressions`. I don't know about the difficulty in doing the check in checkAnalysis. Perhaps, we run into issues while processing the window expression later on in analyzer. 

However, for our current case, i think it makes sense to do it in checkAnalysis since HAVING can have aggregate and is planned as a Filter. I think somewhere in Analyzer we turn this into a Filter over Aggregate with Filter referencing to the output of Aggregate as a named expression. So after we do all that work, when we come to checkAnalysis and we still have aggregate expressions in the Filter node, we raise the error.  Please let me know ..

**Comment by dilipbiswal** [comment]
@cloud-fan For you reference [this pr](https://github.com/apache/spark/pull/21580) has added the check.

**Comment by cloud-fan** [comment]
can you check https://github.com/apache/spark/pull/23658 ?

**Comment by dilipbiswal** [comment]
@cloud-fan Thanks for pointing me to this. I had seen it before but had completely forgotten about it :-). So Wenchen, in my understanding, this PR ensures that the integrity of the plan is not broken in the optimizer between application of rules. Is there something in specific you wanted me to focus on ? One thing i noticed is that here we reject plans which have aggregate expressions and plan node is not Aggregate. In this PR, we check for only Filter .. Did you want to target the fix to other operators as well just like the integrity check pr ? I feel its better to do check specific to problematic operators so we can issue targeted error message. What do you think ?

**Comment by cloud-fan** [comment]
I'm wondering why we still hit this problem after https://github.com/apache/spark/pull/23658

**Comment by dilipbiswal** [comment]
@cloud-fan Oh.. because.. the integrity check is done only in testing mode, right ? So its basically an internal thing ?

**Comment by cloud-fan** [comment]
cc @rednaxelafx @gatorsmile shall we make that check official?

**Comment by gatorsmile** [comment]
I think we should verify it in the analyzer stage. 

The plan integrity verification is just for ensuring the optimizer rules do not change the plan and then break the integrity.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 1]
```diff
@@ -172,16 +176,15 @@ trait CheckAnalysis extends PredicateHelper {
             failAnalysis("Null-aware predicate sub-queries cannot be used in nested " +
               s"conditions: $condition")
 
+          case Filter(condition, _) if condition.find(isAggregateExpression(_)).isDefined =>
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: call plan integrity checking here?

Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: @gatorsmile Thanks for reviewing. I will try it and get back.

Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: @gatorsmile @cloud-fan Just made change to make use of the integrity check method. I had to refactor it out as its a private method in optimizer. Is this what you had in mind ? Please let me know.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 2]
```diff
@@ -172,13 +172,17 @@ trait CheckAnalysis extends PredicateHelper {
             failAnalysis("Null-aware predicate sub-queries cannot be used in nested " +
               s"conditions: $condition")
 
+          case f @ Filter(condition, _) if PlanHelper.specialExpressionInUnsupportedOperator(f) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: why only check filter here? I think we should do it for any operator.

File: sql/core/src/test/resources/sql-tests/inputs/group-by.sql [hunk 3]
```diff
@@ -141,3 +141,11 @@ SELECT every("true");
 SELECT k, v, every(v) OVER (PARTITION BY k ORDER BY v) FROM test_agg;
 SELECT k, v, some(v) OVER (PARTITION BY k ORDER BY v) FROM test_agg;
 SELECT k, v, any(v) OVER (PARTITION BY k ORDER BY v) FROM test_agg;
+
+-- Aggrgate expressions can be referenced through an alias
+SELECT * FROM (SELECT COUNT(*) AS cnt FROM test_agg) WHERE cnt > 1;
+
+-- Error when aggregate expressions are in where clause directly
+SELECT count(*) FROM test_agg WHERE count(*) > 1;
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/group-by.sql]: can we test HAVING as well?

File: sql/core/src/test/resources/sql-tests/results/subquery/negative-cases/invalid-correlation.sql.out [hunk 4]
```diff
@@ -70,9 +70,10 @@ Resolved attribute(s) t2b#x missing from min(t2a)#x,t2c#x in operator !Filter t2
 SELECT t1a 
 FROM   t1
 GROUP  BY 1
-HAVING EXISTS (SELECT 1 
+HAVING EXISTS (SELECT t2a
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/subquery/negative-cases/invalid-correlation.sql.out]: why this change?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 2 above)
Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: @cloud-fan I thought, having a targeted error message for the user would be better. If we do it generically, we would say "Filter" instead of "WHERE clause", right ? Should i have it handled for Filter and have a general catcher as the last case statement to issue a generic error message ?

File: sql/core/src/test/resources/sql-tests/inputs/group-by.sql (hunk 3 above)
Comment (dilipbiswal) [reply sql/core/src/test/resources/sql-tests/inputs/group-by.sql]: @cloud-fan ok.. i will add having.

File: sql/core/src/test/resources/sql-tests/results/subquery/negative-cases/invalid-correlation.sql.out (hunk 4 above)
Comment (dilipbiswal) [reply sql/core/src/test/resources/sql-tests/results/subquery/negative-cases/invalid-correlation.sql.out]: @cloud-fan because, the child subquery plan can not be resolved due to missing attributes. 
```
+- Project [1#81]
:     +- !Filter (t2a#9 < min((outer(t1a#6) + t2a#9))#86)
:        +- Aggregate [1], [1 AS 1#81, min((outer(t1a#6) + t2a#9)) AS min((outer(t1a#6) + t2a#9))#86]
:           +- SubqueryAlias `t2`
:              +- Project [t2a#9, t2b#10, t2c#11]
:                 +- SubqueryAlias `t2`
:                    +- LocalRelation [t2a#9, t2b#10, t2c#11]
````

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: "The operator xxxx contains an unsupported expression type" sounds good

File: sql/core/src/test/resources/sql-tests/results/subquery/negative-cases/invalid-correlation.sql.out (hunk 4 above)
Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/results/subquery/negative-cases/invalid-correlation.sql.out]: But we do expect analysis exception for this query, don't we?

Comment (dilipbiswal) [reply sql/core/src/test/resources/sql-tests/results/subquery/negative-cases/invalid-correlation.sql.out]: @cloud-fan Yeah.. this particular test is in subquery suite and checking for a specific analysis expression that disallows a combination of local and outer references inside an aggregate expression. However, this test had the aggregate expression in the Filter clause which we are disallowing now. So i changed it to use "having" instead of "filter". But when i did that i hit a different analysis exception due to missing attributes (this particular test in question is trying to test a different analysis exception). So i fixed the projection here to include the missing attribute so we get a desired subquery related exception.

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/results/subquery/negative-cases/invalid-correlation.sql.out]: makes sense

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 2 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: do we still need this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/PlanHelper.scala [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+ */
+object PlanHelper {
+  /**
+   * Check if there's any expression in this query plan operator that is
+   * - A WindowExpression but the plan is not Window
+   * - An AggregateExpresion but the plan is not Aggregate or Window
+   * - A Generator but the plan is not Generate
+   * Returns true when this operator hosts illegal expressions. This can happen when
+   * 1. The input query from users contain invalid expressions.
+   *    Example : SELECT * FROM tab WHERE max(c1) > 0
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/PlanHelper.scala]: BTW how about `SELECT max(c1) FROM tab WHERE max(c1) > 0`? Does it work?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 2 above)
Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: @cloud-fan Actually we don't need it. However, since its a user facing message, i feel, its better if we issue targeted error message. I was thinking, we we should add more individual cases as we learn about them here while still having a generic catcher. 
If you think otherwise, then i will remove it. Let me know, please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/PlanHelper.scala (hunk 5 above)
Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/PlanHelper.scala]: @cloud-fan No, this case does not work. But this should work and i have a test case for this.
```
SELECT max FROM (SELECT max(v1) AS max FROM tab) WHERE max > 1
``

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: Then we can move it closer to the general case?

Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: @cloud-fan Sure.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 6]
```diff
@@ -376,6 +376,17 @@ trait CheckAnalysis extends PredicateHelper {
...
+            failAnalysis(
+              s"""
+                 |Aggregate/Window/Generate expressions are not valid in where clause of the query.
+                 |Expression in where clause: ${condition.sql}
+               """.stripMargin)
+
+          case other if PlanHelper.specialExpressionInUnsupportedOperator(other) =>
+            failAnalysis(s"The query operator  `${other.nodeName}` contains " +
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: nit: remove an unnecessary single space.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 7]
```diff
@@ -52,29 +52,7 @@ abstract class Optimizer(sessionCatalog: SessionCatalog)
    * Returns true when all operators are integral.
    */
   private def checkSpecialExpressionIntegrity(plan: LogicalPlan): Boolean = {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: It seems this private function is called by `isPlanIntegral` only, so we need to define this separately?

Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: @maropu Ok... Let me remove it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 6 above)
Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: @maropu Will remove.

**Comment by dilipbiswal** [comment]
@cloud-fan does this look ok now ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/PlanHelper.scala [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.plans.logical
+
+import org.apache.spark.sql.catalyst.expressions.{Generator, WindowExpression}
+import org.apache.spark.sql.catalyst.expressions.aggregate.AggregateExpression
+
+/**
+ * [[PlanHelper]] Contains utility methods that can be used by Analyzer and Optimizer.
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/PlanHelper.scala]: super nit: Contains -> contains.

Don't need to wait for another round test for this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/PlanHelper.scala [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+ * It can also be container of methods that are common across multiple rules in Analyzer
+ * and optimizer.
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/PlanHelper.scala]: super nit: optimizer -> Optimizer. As consistent to above line.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/PlanHelper.scala [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+   * Check if there's any expression in this query plan operator that is
+   * - A WindowExpression but the plan is not Window
+   * - An AggregateExpresion but the plan is not Aggregate or Window
+   * - A Generator but the plan is not Generate
+   * Returns true when this operator hosts illegal expressions. This can happen when
+   * 1. The input query from users contain invalid expressions.
+   *    Example : SELECT * FROM tab WHERE max(c1) > 0
+   * 2. Query rewrites inadvertently produce plans that are invalid.
+   */
+  def specialExpressionInUnsupportedOperator(plan: LogicalPlan): Boolean = {
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/PlanHelper.scala]: what if we return the unsupported operator here and we report the specific one which is problematic to the user, rather than printing the whole filter condition? I think this would help the user to understand which specific thing caused the issue and it would be more user friendly.

Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/PlanHelper.scala]: @mgaido91 The caller already knows about the problematic operator, no ? About the problematic expression, i suppose we could return the first occurence of the problematic expression and report it in the error text.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 11]
```diff
@@ -376,6 +376,28 @@ trait CheckAnalysis extends PredicateHelper {
...
+
+          case other if PlanHelper.specialExpressionInUnsupportedOperator(other).nonEmpty =>
+            val invalidExprSqls =
+              PlanHelper.specialExpressionInUnsupportedOperator(other).map(_.sql)
+            failAnalysis(
+              s"""
+                 |The query operator `${other.nodeName}` contains one or more unsupported
+                 |expression types Aggregate, Window or Generate.
+                 |Invalid expressions: [${invalidExprSqls.mkString(", ")}]
+                 |
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: nit: can we remove this line? and also close the string on the line above, in order to avoid weird things like those `;` in the middle of nowhere in the output

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/PlanHelper.scala [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+   * Check if there's any expression in this query plan operator that is
+   * - A WindowExpression but the plan is not Window
+   * - An AggregateExpresion but the plan is not Aggregate or Window
+   * - A Generator but the plan is not Generate
+   * Returns the list of invalid expressions that this operator hosts. This can happen when
+   * 1. The input query from users contain invalid expressions.
+   *    Example : SELECT * FROM tab WHERE max(c1) > 0
+   * 2. Query rewrites inadvertently produce plans that are invalid.
+   */
+  def specialExpressionInUnsupportedOperator(plan: LogicalPlan): Seq[Expression] = {
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/PlanHelper.scala]: nit: `specialExpressions`?

**Review by mgaido91 — COMMENTED** [review]
only few minor comments, otherwise LGTM, thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 11 above)
Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: @mgaido91 thanks.. will remove. By closing on the line above you mean like ..
```
|Invalid expressions: [${invalidExprSqls.mkString(", ")}]""".stripMargin
```
?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/PlanHelper.scala (hunk 12 above)
Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/PlanHelper.scala]: @mgaido91 Will change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 11 above)
Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: yes, I meant that

**Comment by dilipbiswal** [comment]
@cloud-fan Please let me know if this looks okay now ?

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by dilipbiswal** [comment]
@cloud-fan Thank you very much !!
